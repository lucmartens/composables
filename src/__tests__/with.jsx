import React from 'react';
import { mount } from 'enzyme';

import With from '../with';

describe('With calls', () => {
  test('enter on mount with input', () => {
    const enter = jest.fn();
    const wrapper = mount(<With input="a" enter={enter} />);

    expect(enter.mock.calls.length).toBe(1);
    expect(enter.mock.calls[0]).toMatchSnapshot();
  });

  test('render with output', () => {
    const exit = jest.fn();
    const enter = jest.fn(input => input + 'b');
    const render = jest.fn(output => 'a');
    const wrapper = mount(
      <With input="a" enter={enter} exit={exit} render={render} />
    );

    expect(render.mock.calls.length).toBe(1);
    expect(render.mock.calls[0]).toMatchSnapshot();
  });

  test('render with uninitialized output when lazy', () => {
    const exit = jest.fn();
    const enter = jest.fn(input => input + 'b');
    const render = jest.fn(output => 'a');
    const wrapper = mount(
      <With input="a" enter={enter} exit={exit} render={render} lazy={true} />
    );

    expect(render.mock.calls.length).toBe(2);
    expect(render.mock.calls[0]).toMatchSnapshot();
    expect(render.mock.calls[1]).toMatchSnapshot();
  });

  test('exit on unmount with output and input', () => {
    const exit = jest.fn();
    const enter = jest.fn(input => input + 'b');
    const wrapper = mount(<With input="a" enter={enter} exit={exit} />);

    expect(exit).not.toBeCalled();
    wrapper.unmount();
    expect(exit.mock.calls.length).toBe(1);
    expect(exit.mock.calls[0]).toMatchSnapshot();
  });

  test('shouldUpdate on input change with previous and next input', () => {
    const fn = jest.fn();
    const enter = jest.fn(input => input + 'b');
    const wrapper = mount(<With input="a" enter={enter} shouldUpdate={fn} />);

    expect(fn).not.toBeCalled();
    wrapper.setProps({ input: 'b' });
    expect(fn.mock.calls.length).toBe(1);
    expect(fn.mock.calls[0]).toMatchSnapshot();
  });

  test('exit on input change', () => {
    const exit = jest.fn();
    const enter = jest.fn(input => input + 'b');

    const wrapper = mount(<With input="a" enter={enter} exit={exit} />);

    expect(exit.mock.calls.length).toBe(0);
    wrapper.setProps({ input: 'b' });
    expect(exit.mock.calls.length).toBe(1);
    expect(exit.mock.calls[0]).toMatchSnapshot();
  });

  test('enter on input change', () => {
    const exit = jest.fn();
    const enter = jest.fn(input => input + 'b');

    const wrapper = mount(<With input="a" enter={enter} exit={exit} />);

    expect(enter.mock.calls.length).toBe(1);
    wrapper.setProps({ input: 'b' });
    expect(enter.mock.calls.length).toBe(2);
    expect(enter.mock.calls[1]).toMatchSnapshot();
  });
});
