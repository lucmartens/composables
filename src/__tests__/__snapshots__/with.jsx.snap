// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`With calls enter on input change 1`] = `
Array [
  "b",
]
`;

exports[`With calls enter on mount with input 1`] = `
Array [
  "a",
]
`;

exports[`With calls exit on input change 1`] = `
Array [
  "ab",
  "a",
]
`;

exports[`With calls exit on unmount with output and input 1`] = `
Array [
  "ab",
  "a",
]
`;

exports[`With calls render with output 1`] = `
Array [
  "ab",
]
`;

exports[`With calls render with uninitialized output when lazy 1`] = `
Array [
  undefined,
]
`;

exports[`With calls render with uninitialized output when lazy 2`] = `
Array [
  "ab",
]
`;

exports[`With calls shouldUpdate on input change with previous and next input 1`] = `
Array [
  "a",
  "b",
]
`;
